swupd/manifest.go

```diff
@@ -363,59 +363,100 @@ func (m *Manifest) prependFile(f *File) {
 // At this point Manifest m should only have the files that were present
 // in the chroot for that manifest. Link delta peers with the oldManifest
 // if the file in the oldManifest is not deleted or ghosted
-func (m *Manifest) linkPeersAndChange(oldManifest *Manifest) int {
+func (m *Manifest) linkPeersAndChange(oldManifest *Manifest) (int, int, int) {
+	m.sortFilesName()
+	oldManifest.sortFilesName()
+
 	changed := 0
-	for _, of := range oldManifest.Files {
-		if match := of.findFileNameInSlice(m.Files); match != nil {
+	added := 0
+	deleted := 0
+	i := 0
+	j := 0
+	for i < len(m.Files) && j < len(oldManifest.Files) {
+		nf := m.Files[i]
+		of := oldManifest.Files[j]
+		if nf.Name == of.Name {
+			// if it is the same name check if anything about the file has changed.
+			// if something has changed update the version to current version and
+			// record that the file has changed
 			if of.Status == statusDeleted || of.Status == statusGhosted {
+				// don't worry about ghosted or deleted files
+				i++
+				j++
 				continue
 			}
 
-			match.DeltaPeer = of
-			of.DeltaPeer = match
-			if !sameFile(match, of) {
-				match.Version = m.Header.Version
+			// set up peers
+			nf.DeltaPeer = of
+			of.DeltaPeer = nf
+			if !sameFile(nf, of) {
+				// if the file isn't exactly the same, record the change
+				// and update the version
+				nf.Version = m.Header.Version
 				changed++
 			} else {
-				match.Version = of.Version
+				// otherwise the version is the same as the old file
+				nf.Version = of.Version
+			}
+			// advance indices for both file lists since we had a match
+			i++
+			j++
+		} else if nf.Name < of.Name {
+			// if the file does not exist in the old manifest it is a new
```

Comment:
Don't we want at this stage to add the nf to a list of potential new files, and then the rename detection after this loop?

Reply:
Well, it isn't just new files that could be a renamed-to file if the rename is from an older version, but I suppose the rename status will be copied over from the old manifest in that case. Alternatively we can just run rename detection over the manifest files for files that are last changed at the current version.